### ISSUE TITLE
Cannot use relative path following placeholder in @TestPropertySource locations

### ISSUE THREAD

**Issue by Pancor** [spring-projects/spring-framework#23544]
**Affects:** Spring Boot version: 2.1.1.RELEASE with Spring-test version: 5.1.3.RELEASE

I'm trying to add custom path to `application.properties` file for one of my class test with use of `@TestPropertySource` like this:

```java
@RunWith(SpringRunner.class)
@SpringBootTest
@TestPropertySource(locations = "file:${user.dir}/../config/application.properties")
public class AuthControllerITest {
    //some code
}
```

And I have an error caused by `FileNotFoundException` which says that it can not find this path: `config/application.properties`. When I remove this ".." characters (so i've got `file:${user.dir}/config/application.properties`), then it gives me this path:  `/home/pawel/projekty/simulatorserver/server/config/application.properties`, but I want to get to this path: `/home/pawel/projekty/simulatorserver/config/application.properties` (without server directory).

I did some investigation and I found out that [org.springframework.util.StringUtils#cleanPath(String path)](https://github.com/spring-projects/spring-framework/blob/8d86a861a1eff13dd1abe14829e4213200509300/spring-core/src/main/java/org/springframework/util/StringUtils.java#L648) removes ${user.dir} from path. 

I guess it should first resolve ${user.dir} to path, and then remove one directory from it.

For now I'm satisfied with this solution:

```java
@RunWith(SpringRunner.class)
@SpringBootTest
@TestPropertySource(locations = "file:../config/application.properties")
public class AuthControllerITest {
    //some code
}
```

which gives me proper output: `/home/pawel/projekty/simulatorserver/config/application.properties`.

Stacktrace:
```
java.lang.IllegalStateException: Failed to load ApplicationContext

	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:125)
	at org.springframework.test.context.support.DefaultTestContext.getApplicationContext(DefaultTestContext.java:108)
	at org.springframework.test.context.web.ServletTestExecutionListener.setUpRequestContextIfNecessary(ServletTestExecutionListener.java:190)
	at org.springframework.test.context.web.ServletTestExecutionListener.prepareTestInstance(ServletTestExecutionListener.java:132)
	at org.springframework.test.context.TestContextManager.prepareTestInstance(TestContextManager.java:246)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.createTest(SpringJUnit4ClassRunner.java:227)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner$1.runReflectiveCall(SpringJUnit4ClassRunner.java:289)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.methodBlock(SpringJUnit4ClassRunner.java:291)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:246)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:97)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:190)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: java.lang.IllegalStateException: Failed to add PropertySource to Environment
	at org.springframework.test.context.support.TestPropertySourceUtils.addPropertiesFilesToEnvironment(TestPropertySourceUtils.java:201)
	at org.springframework.boot.test.context.SpringBootContextLoader.loadContext(SpringBootContextLoader.java:104)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContextInternal(DefaultCacheAwareContextLoaderDelegate.java:99)
	at org.springframework.test.context.cache.DefaultCacheAwareContextLoaderDelegate.loadContext(DefaultCacheAwareContextLoaderDelegate.java:117)
	... 24 more
Caused by: java.io.FileNotFoundException: config/application.properties (No such file or directory)
	at java.base/java.io.FileInputStream.open0(Native Method)
	at java.base/java.io.FileInputStream.open(FileInputStream.java:219)
	at java.base/java.io.FileInputStream.<init>(FileInputStream.java:157)
	at java.base/java.io.FileInputStream.<init>(FileInputStream.java:112)
	at java.base/sun.net.www.protocol.file.FileURLConnection.connect(FileURLConnection.java:86)
	at java.base/sun.net.www.protocol.file.FileURLConnection.getInputStream(FileURLConnection.java:184)
	at org.springframework.core.io.UrlResource.getInputStream(UrlResource.java:173)
	at org.springframework.core.io.support.EncodedResource.getInputStream(EncodedResource.java:159)
	at org.springframework.core.io.support.PropertiesLoaderUtils.fillProperties(PropertiesLoaderUtils.java:99)
	at org.springframework.core.io.support.PropertiesLoaderUtils.fillProperties(PropertiesLoaderUtils.java:73)
	at org.springframework.core.io.support.PropertiesLoaderUtils.loadProperties(PropertiesLoaderUtils.java:59)
	at org.springframework.core.io.support.ResourcePropertySource.<init>(ResourcePropertySource.java:86)
	at org.springframework.test.context.support.TestPropertySourceUtils.addPropertiesFilesToEnvironment(TestPropertySourceUtils.java:197)
	... 27 more
```

**Comment by sbrannen** [comment]
Thanks for raising the issue.

It's certainly a unique problem that only appears to manifest itself if you have a placeholder immediately followed by `/../` in the supplied path.

> `@TestPropertySource(locations = "file:../config/application.properties")`

That is indeed the best way to reference the parent directory for the currently executing project. So I'd recommend sticking with that.

As for the general case where a placeholder is followed by `/../` in the supplied path, I'm not sure that we can "fix" that without introducing a breaking change, but we'll investigate our options.

**Comment by sbrannen** [comment]
This has been fixed in `master` in ab779eb431ebcfaf9c13b36cac905789c12b610c.

Feel free to try it out in the next 5.2 snapshot build.

**Comment by Pancor** [comment]
Thanks, great to hear that :)
